New introductory flow (#82)
sig-rpc/sig-rpc.github.io: pages/getting-started.md

<!-- Could add mathjax to render the formula $O = \frac{1}{(1-P) + \frac{P}{S}}$ -->
To quantify this idea, we can use Amdahl's law which shows that the overall performance improvement from optimising a single part of a system is limited by the fraction of time that the improved part is actually used. [Figure 1](#interactive-amdahls) provides an interactive graph where you can explore changing the the proportion of runtime optimised (`P`).

{% figure caption:"An interactive graph demonstrating Amdahl's law, the overall speedup following a speedup (S) to propotion (P) of the code." label:interactive-amdahls %}
<!-- Interactive Amdahl's law graph, vibe coded with GPT -->
<!-- @todo can I wrap this inside a captioned Jekyll figure? -->
<canvas id="graph" width="600" height="400"></canvas>

<div class="controls">
  <label><tiny>
    Proportion of runtime optimised (P):
    <span id="pValue">0.20</span>
  </tiny></label>
  <input type="range" id="pSlider" min="0" max="100" value="20">
</div>
{% endfigure %}

<!-- Not sure I like this -->

> The above graph, clearly demonstrates how theoretically optimising away 20% of the runtime to 0% would provide a 1.25x speedup, showing why we should focus optimisation attention on the significant components. There's little value in optimising away 1% of the runtime if it will have negligible impact to performance.

Reviewer: ```suggestion
The above graph theoretically demonstrates how optimising so as to remove 20% of the runtime would provide a 1.25x speedup, showing why we should focus optimisation attention on the significant components. There's little value in optimising away 1% of the runtime if it will have negligible impact to performance (in this example only a 1.01x speedup).
```

<!-- Could link to FFEA case-study at final sentence here -->
However Amdahl's law is theoretical, in practice computer hardware is shared and bottlenecks may impact multiple components of the code. Hence, it's possible for some optimisations to improve performance beyond the modified code. A memorable example, saw a profile attribute slightly below 80% of the runtime to redundant object copies, when removed this led to a 10x speedup, yet Amdahl's law would predict a 5x speedup.

You may have previously tried to identify performance bottlenecks by manually timing sections of your code, for example by inserting timing calls around individual functions or blocks. This approach is slow and unwieldy. It is difficult to achieve good coverage even for a moderately sized codebase, and the added instrumentation often requires ongoing effort to update, maintain, and eventually remove as a project evolves.

> In contrast, the most basic form of profiling, hot-path profiling (also known as hotspot profiling), requires little additional expertise if you understand your code and can usually be integrated quickly into your release workflow. Profiling may immediately reveal an obvious bottleneck, confirm your assumptions about the most expensive parts of your code, or highlight areas that need further investigation. It is a low-cost exercise that either leads to tangible improvements or gives you greater confidence in the performance of your code.

Reviewer: ```suggestion
In contrast, the most basic form of profiling, hot-path profiling (also known as hotspot profiling), requires little additional expertise if you understand your code and can usually be integrated quickly into your workflow. Profiling may immediately reveal an obvious bottleneck, confirm your assumptions about the most expensive parts of your code, or highlight areas that need further investigation. It is a low-cost exercise that either leads to tangible improvements or gives you greater confidence in the performance of your code.
```
Plenty of people won't be thinking about this in the context of releases

It is not essential to memorise these differences. In most cases, you can simply use the profiler that is available. The key points are summarised in [figure 2](#profiler-type-table) below.

{% figure caption:"A table summarsing the strengths and weaknesses of different profiling tools." label:profiler-type-table %}
| Collection method | Attribution level | Strengths                           | Weaknesses       |
| ----------------- | ----------------- | ----------------------------------- | ----------------- |
| Sampling          | Function          | Minimal overhead                    | Statistical noise for small runs |
| Sampling          | Line              | Low overhead, fine-grained insight  | Potentially biased by timing  |
| Deterministic     | Function          | Exact call counts, timings          | Moderate overhead |
| Deterministic     | Line              | Precise attribution, branching info | High overhead     |
{% endfigure %}

In addition to common hot-path profiling tools, you may come across more specialist profilers. In general these take two forms.

> Firstly there are timeline profilers, these typically operate similar to function-level profilers however instead of presenting aggregations of the function-level metrics, they present function calls on a timeline of the code's execution. Timeline profiling is particularly useful when investigating parallel codes, as it clearly displays synchronisation points and idle threads. It may also be useful if profiling a code, where performance degrades over time as this can be difficult to diagnose with aggregated metrics.

Reviewer: ```suggestion
Firstly there are timeline profilers, these typically operate similar to function-level profilers however instead of presenting aggregations of the function-level metrics, they present function calls on a timeline of the code's execution. Timeline profiling is particularly useful when investigating parallel codes as it clearly displays synchronisation points and idle threads. It may also be useful if profiling a code where performance degrades over runtime as this can be difficult to diagnose with aggregated metrics.
```
